pkg/azuredisk/controllerserver.go

```diff
@@ -517,60 +525,170 @@ func (d *Driver) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
 			return nil, status.Errorf(codes.Aborted, "ListVolumes starting token(%d) can not be negative", start)
 		}
 	}
+	if d.cloud.KubeClient != nil && d.cloud.KubeClient.CoreV1() != nil && d.cloud.KubeClient.CoreV1().PersistentVolumes() != nil {
+		klog.V(6).Infof("List Volumes in Cluster:")
+		return d.listVolumesInCluster(ctx, start, int(req.MaxEntries))
+	}
+	klog.V(6).Infof("List Volumes in Node Resource Group:")
+	return d.listVolumesInNodeResourceGroup(ctx, start, int(req.MaxEntries))
+}
 
-	disks, derr := d.cloud.DisksClient.ListByResourceGroup(ctx, d.cloud.ResourceGroup)
-	if derr != nil {
-		return nil, status.Error(codes.Internal, fmt.Sprintf("ListVolumes on rg(%s) failed with error: %v", d.cloud.ResourceGroup, derr.Error()))
+// listVolumesInCluster is a helper function for ListVolumes used for when there is an available kubeclient
+func (d *Driver) listVolumesInCluster(ctx context.Context, start, maxEntries int) (*csi.ListVolumesResponse, error) {
+	pvList, err := d.cloud.KubeClient.CoreV1().PersistentVolumes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "ListVolumes failed while fetching PersistentVolumes List with error: %v", err.Error())
+	}
+
+	// get all resource groups and put them into a sorted slice
+	rgMap := make(map[string]bool)
+	volSet := make(map[string]bool)
+	for _, pv := range pvList.Items {
+		if pv.Spec.CSI != nil && pv.Spec.CSI.Driver == DriverName {
+			diskURI := pv.Spec.CSI.VolumeHandle
+			if err := isValidDiskURI(diskURI); err != nil {
+				klog.Warningf("invalid disk uri (%s) with error(%v)", diskURI, err)
+				continue
+			}
+			rg, err := GetResourceGroupFromURI(diskURI)
+			if err != nil {
+				klog.Warningf("failed to get resource group from disk uri (%s) with error(%v)", diskURI, err)
+				continue
+			}
+			volSet[diskURI] = true
+			if _, visited := rgMap[rg]; visited {
+				continue
+			}
+			rgMap[rg] = true
+		}
 	}
 
-	if start != 0 && start >= len(disks) {
+	if start != 0 && start > len(volSet) {
```

Comment:
start token refers to an index in a list of all azure disks, including those that may not be csi volumes. This condition compares the start token index to the number of azure csi volumes; therefore it's an invalid comparison.